src/databricks/sql/utils.py

```diff
@@ -470,3 +474,66 @@ def _create_arrow_array(t_col_value_wrapper, arrow_type):
             result[i] = None
 
     return pyarrow.array(result, type=arrow_type)
+
+class DbSqlType(Enum):
+    STRING = "STRING"
+    DATE = "DATE"
+    TIMESTAMP = "TIMESTAMP"
+    FLOAT = "FLOAT"
+    DECIMAL = "DECIMAL"
+    INTEGER = "INTEGER"
+    BIGINT = "BIGINT"
+    SMALLINT = "SMALLINT"
+    TINYINT = "TINYINT"
+    BOOLEAN = "BOOLEAN"
+    INTERVAL_MONTH = "INTERVAL MONTH"
+    INTERVAL_DAY = "INTERVAL DAY"
+
+class DbSqlParameter:
+    name: str
+    value: Any
+    type: DbSqlType
+    def __init__(self,name="",value=None,type=None):
+        self.name = name
+        self.value = value
+        self.type = type
+        
+        
+def named_parameters_to_dbsqlparams(parameters: Any):
+    dbsqlparams = []
+    if(type(parameters) is dict):
+        # Support legacy impl
+        for name,parameter in parameters.items():
+            dbsqlparams.append(DbSqlParameter(name=name,value=parameter))
+    else:
+        for parameter in parameters:
+            if(type(parameter) is DbSqlParameter):
+                dbsqlparams.append(parameter)
+            else:
+                dbsqlparams.append(DbSqlParameter(value=parameter))
+    return dbsqlparams
+
+def infer_types(params: list[DbSqlParameter]):
```

Comment:
We should prefer [pure functions](https://pythoninformer.com/programming-techniques/functional-programming/pure-functions/) wherever possible. `infer_types` isn't pure right now which makes it harder to test and therefore more likely to produce bugs.

It should be rewritten to keep its inputs immutable (don't modify their values in place) and return a new list.